### ISSUE TITLE
Run riboviz on specific nodes on Eddie, to solve memory allocation problem

### ISSUE THREAD

**Issue by ewallace** [riboviz/riboviz#236]
I wrote to the ECDF Eddie team 17/11/20 with the following explanation:

> We would like to ask for help on the memory requirements for job submission on Eddie. Basically, we are running bioinformatics workflows using the nextflow environment, and we have been struggling with jobs getting killed due to running over their memory allocation. I am trying to solve this by requesting an entire standard 64GB node, using `-pe mpi 16 -l h_vmem 4GB -R y`. But then sometimes the job gets allocated to (I think) a standard node with more than 64GB (16x4GB) memory, then the job requests more memory on the node, and gets killed.
> 
> Can we request job submission only on standard 64GB nodes? If it is easier, we would be fine for the foreseeable future if we restricted the bio_wallace_rna_riboviz queue to only access the standard 64GB nodes.
> 
> If there is a way to restrict to certain nodes as a qsub option, that would be great to put on the Eddie wiki.
> 
> For context: we have no way that we know of to restrict the amount of memory that the individual steps in the pipeline request. We are running a whole lot of 3rd-party bioinformatics software that each has its own ways of requesting resources. So it's probably easiest to run the workflow in a place where the steps can't request more than they "see" - i.e. an entire node. Hence my suggestion.

Mike Wallis (head of Eddie) replied on 24/11/20 with the following helpful suggestions:

> This is an interesting one; there's a more elegant way of doing it which I'd like to implement but the quick and dirty hack which should get you running is to limit the hosts you submit to.
> 
> Most of the 16/64G nodes are in two specific node ranges, node1b* and 1c* - so you could submit your jobs using a hostname complex:
> 
> # h = hostname regexp limiting to nodes only in 1b and 1c
> #$ -l h='(node1b*|node1c*)'
> 
> The more elegant solution would be to have a specific complex set up for cpu/ram combinations, which allows the concept of "islands" - once they're set up we can configure your project to run on those islands. This might take a bit of time, though.
> 
> Can you let me know how you get on with the hostname regexp in the first instance, please?

We agreed at 2/12/20 Dev meeting that @FlicAnderson will try the hack the next time she runs a pipeline on Eddie in the next week. We can assess whether that works.

In the longer term we will be in touch with the Eddie team and try their "islands" solution when it is set up.

**Comment by ewallace** [comment]
Mike Wallis just responded:
> We've implemented a slightly better solution for you. Can you try doing
> `qsub -q *@@uoe_64G <jobscript>`
> instead, please?

@FlicAnderson do you have time for that?

**Comment by FlicAnderson** [comment]
I've tried running some data with [Mike Wallis' new solution](https://github.com/riboviz/riboviz/issues/236#issuecomment-738064790) and the job is in the queue, but also not sure whether it makes any of the other Grid Engine parameters redundant or not?  @ewallace is this something you could ask Mike Wallis?  Also what this fix is, and what it does?  I'm not sure what logic to apply when I'm trying to set the other GE options otherwise.

**Comment by ewallace** [comment]
@FlicAnderson my understanding is that`-q *@@uoe_64G` restricts to nodes with 16 processors and 64GB RAM. So it should be combined with `-pe mpi 16 -l h_vmem=4GB` to ensure that you are requesting 64GB RAM and a whole node. There may be other combinations that work, and I may be wrong.

It will be great to test, document, then close.

**Comment by FlicAnderson** [comment]
@ewallace I've submitted a version with those (`-pe mpi 16 -l h_vmem=4GB`) params. Will update.

**Comment by FlicAnderson** [comment]
(Reminder @ewallace is relying on my results of this run, so I can update here and allow him to proceed with that)

**Comment by FlicAnderson** [comment]
Job began running reasonably quickly, but aborted very quickly too...

```
# B-Sc_2012 dataset run attempt on Eddie

# submitted via: qsub -q *@@uoe_64G <jobscript> method
# selected Grid Engine options: 
#$ -l h_vmem=4G 
#$ -pe mpi 16

# summary output from the job submission email:
Job 9021377 (B-Sc_2012) Aborted
 Exit Status      = 137
 Signal           = KILL
 User             = fanders6
 Queue            = eddie@node1c10.ecdf.ed.ac.uk
 Host             = node1c10.ecdf.ed.ac.uk
 Start Time       = 12/16/2020 16:55:30.469
 End Time         = 12/16/2020 17:20:09.139
 CPU              = 03:56:44
 Max vmem         = 67.990G
 Max rss          = NA
failed execd enforced h_vmem limit because:
job 9021377.1 died through signal KILL (9)
```
Not enough memory :sob:

**Comment by ewallace** [comment]
Huh. That is **exactly** the problem we were trying to avoid. I'll ask @mikewallis to advise. I wonder if node1c10 was a 64G node, or not?

**Comment by ewallace** [comment]
Angela Chitzanidi from ECDF has helped us with a workaround, consisting of a couple of nodes with no virtual memory restrictions.

They are two 

> 40-core nodes with 768GB RAM each. I've set them up so you can request as much virtual memory as you like, whether in one or multiple slots (they have 40 slots per node).

You 

> can submit to these nodes as follows:

> 
> `qsub  -pe sharedmem N -l h_vmem=MG -q *@@ringfence_wallace jobscript`
> 
> where:
> 
> - N is the number of slots (maximum 40)
> - M is the amount of virtual memory per slot (as much as you like - I would recommend choosing a suitably high number so that you don't run into the same issues again)
> - jobscript the name of your job submission script file.
> 
> Once you've done a successful run please let me know providing the job id so we can check how much virtual memory these jobs actually do try to access.

These should be accessible to users in `wallace_rna` group. In January, let's make a plan to try these. We should tally the virtual memory used with the nextflow report, #231.

**Comment by FlicAnderson** [comment]
Latest findings from testing Angela's latest incantation involving ringfencing (sent by email pre-Christmas) on a full-size Drosophila dataset (example-datasets/#10) A-Dm_2014:

Angela's email:

> I've set up these nodes now. They are two 40-core nodes with 768GB RAM each.
> I've set them up so you can request as much virtual memory as you like, whether in one or multiple slots (they have 40 slots per node).
> You can submit to these nodes as follows:
> `qsub  -pe sharedmem N -l h_vmem=MG -q *@@ringfence_wallace jobscript`
> where:
> - N is the number of slots (maximum 40)
> - M is the amount of virtual memory per slot (as much as you like - I would recommend choosing a suitably high number so that you don't run into the same issues again)
> - jobscript the name of your job submission script file.
> Once you've done a successful run please let me know providing the job id so we can check how much virtual memory these jobs actually do try to access.

I submitted the job via: 
```
$ qsub -pe sharedmem 32 -l h_vmem=256G -q *@@ringfence_wallace run_A-Dm_2014_fullsize.sh 
```
Submitted: 21:04:46
Started: 21:05:00
SUCCESS! Began running very quickly indeed.

It also completed fine, although I definitely think h_vmem=256G was waaaay more than required for this job, as the max vmem output in the job completion info was 75.287G.  It might be good to see if this can be reduced to a more reasonable level without upsetting anything.  I'm also not sure if 32 was too high for number of slots, but I started high as it had higher chances of completion success :grin: 

I would like to test a few other datasets via this particular incantation, and then we can report back to Angela & Mike Wallis and get their opinions on how to proceed in future.

TL;DR: the ringfenced submission command runs ASAP, it's great!
(comment moved from #230 where it didn't belong.)

**Comment by FlicAnderson** [comment]
It would be good to contact Angela & Mike again regarding whether this `qsub -pe sharedmem N -l h_vmem=MG -q *@@ringfence_wallace jobscript` ringfenced submission is an OK mid/long term solution, since it works really really well for submitting jobs just now.

**Comment by FlicAnderson** [comment]
> It would be good to contact Angela & Mike again regarding whether this `qsub -pe sharedmem N -l h_vmem=MG -q *@@ringfence_wallace jobscript` ringfenced submission is an OK mid/long term solution, since it works really really well for submitting jobs just now.

Emailed Angela just now enquiring what information would be useful to figure out the memory usage for our dataset runs, so I can compile this and move this issue forward, and also asked whether it's still ok to keep submitting using this ringfence formula, as it's working amazingly :grinning:  If it IS ok to keep submitting in this way, it's probably worth editing #228 to follow this format of job submission command, as it means jobs start almost instantly rather than not starting, which seemed to be the case for a job submission script I recently generated using the code in #228's branch.

**Comment by FlicAnderson** [comment]
Today (24 Mar 2021), I emailed Angela with a list of job-ids from Eddie runs to help figure out the resource allocation situation: 

> Hi Angela,
> Here are some 'typical' jobs we've run through riboviz workflow using the restricted queue:
>
> Format:  job-id (job name for my reference)
>
> - 10759689 (L-Sc_2014)
> - 9932373 (S-Mm_2017)
> - 9838315 (W-Sc_2016)
> - 9973456 (M-Ca_2014)
> 
> These last two below are examples of running non-workflow tools. 
> - 9920713 (gff_fa_A-Dm_2014)
> - 9935183 (SRA_M-Ca_2014)
> 
> Hope this is helpful in establishing the resource use. If you need more information, please let me know.

This should let her follow up to identify what's needed, as per her request: 
> I have accounting information going back a year, so if you just provide  job ids I would be able to extract any information we need.
> With regards to costs, for now you have been given this node for free for testing, but once we confirm how much resource you actually require, we can arrange for a paid ringfence for you for a set period of time.
> As a first step, if you provide some job ids so  I can work out how much resource it is you have been actually using, that would be good.

I'll feed back on the ticket with her reply.

**Comment by FlicAnderson** [comment]
This is the latest reply from Angela: 

She's figured out the optimum job submission specs which we should use and removed us from the temporary ringfenced node. 

> Hi Felicity,
> Here is the relevant accounting information from the jobs you sent me:
> 
> Job 10759689
> wallclock    8207.903
> cpu          76652.258
> maxvmem      131.088G
> 
> Job 9932373
> wallclock    74633.223
> cpu          592190.899
> maxvmem      175.383G
> 
> Job 9838315
> wallclock    21446.106
> cpu          58168.538
> maxvmem      71.823G
> 
> Job 9973456
> wallclock    14420.106
> cpu          96740.813
> maxvmem      73.729G
> 
> Job 9920713
> wallclock    686.064
> cpu          672.308
> maxvmem      327.492M
> 
> Job 9935183
> wallclock    1374.896
> cpu          6220.481
> maxvmem      2.724G
> 
> Dividing the 'cpu' value by the 'wallclock' value will give you the amount of cores the job used. The maximum virtual memory it tried to access can be seen from the 'maxvmem' value.
> 
> From the above, the maximum values for both core number and memory are:
> 
> Maximum virtual memory used: 175.383GB (job 9932373)
> Maximum number of cores used: 9 cores  (job 10759689)
> 
> It is clear that you don't need to be submitting your jobs with a 32-slot request and you don't need a total of more than 200GB of RAM per job either. Depending on the size of the  job, you should be submitting them with a slot number and memory per slot request that falls within the figures above. For the largest size of job you have run so far that would be something like
> 
> #$ -pe  sharedmem 9
> #$ -l h_vmem=20G
> 
> This would give you 9 cores and a total of 180GB of RAM.
> 
> In terms of setting up a compute project, you have two options:
> 
> 1) You can have a priority project which doesn't give you exclusive access to any particular node, but gives you higher priority in the queue.
> 2) You can have a ringfenced node which will give you exclusive access like the arrangement we have had so far for testing. Note that this is much more expensive though, because you pay  for the ringfenced resource regardless of whether you use it or not.
> 
> Either way, it is clear that you can fit at least 3 or 4 jobs at the same time on one of our larger nodes (32-core or 40-core).

So these `sharedmem`/`h_vmem` values Angela has picked out should be really useful for using to run on Eddie from now on, and should probably be used by me, @ewallace, @3mma-mack and @swinterbourne.  (Also pinging @mikej888 in case this is good to know for anything you're working on in future that may need testing with a large dataset?)

**Comment by kavousan** [comment]
Updates on 30 June 2021.

* We probably want to be using the priority queue. For that, please specify: `-P bio_wallace_rna_riboviz`
* **If** (and it's a bif **if**) we want to be using exclusive nodes for some crucial runs, then Angela recommends "aiming to land on 192GB nodes with  `-pe sharedmem 16 -l h_vmem=12G -q *@@uoe_192G`."
